### DISCUSSION TITLE
Is there a way to forward declaration in the header?

### DISCUSSION THREAD

**Discussion by longod** [simdjson/simdjson#2097] [Q&A]
Hi. Is there a way to forward declaration in the header?
Is it better to use a template function or something else?

For example, I want to declare an API that uses `simdjson::ondemand::value` as an argument.

This is an error because the namespace differs from the actual definition.
```cpp
namespace simdjson {
namespace ondemand {
class value;
}
}
bool do_something(simdjson::ondemand::value& value);
```

This is interpreted as a namespace where `SIMDJSON_IMPLEMENTATION` is not a define.
```cpp
namespace simdjson {
namespace SIMDJSON_IMPLEMENTATION {
namespace ondemand {
class value;
}
}
}
bool do_something(simdjson::SIMDJSON_IMPLEMENTATION::ondemand::value& value);
```

Enumerating by all architectures may be ok, but hard to maintain.
```cpp
namespace simdjson {
namespace arm64 {
namespace ondemand {
class value;
}
}
namespace fallback {
namespace ondemand {
class value;
}
}
namespace haswell {
namespace ondemand {
class value;
}
}
namespace icelake {
namespace ondemand {
class value;
}
}
namespace ppc64 {
namespace ondemand {
class value;
}
}
namespace westmere {
namespace ondemand {
class value;
}
}
}

bool do_something(simdjson::arm64::ondemand::value& value);
bool do_something(simdjson::fallback::ondemand::value& value);
bool do_something(simdjson::haswell::ondemand::value& value);
bool do_something(simdjson::icelake::ondemand::value& value);
bool do_something(simdjson::ppc64::ondemand::value& value);
bool do_something(simdjson::westmere::ondemand::value& value);
```

**Comment by lemire** [comment]
There should be no problem passing `simdjson::ondemand::value value` to a function...


```C++
#include "simdjson.h"
#include <iostream>


void f(simdjson::ondemand::value v) {
  double val;
  if(auto error = v.get_double().get(val)) { std::cout << error << std::endl; return; }
  printf("%.0f\n", val);
  std::cout << std::hexfloat << val << std::endl;
}

int main(void) {
  simdjson::padded_string json = R"( [ 233897314173811950000 ])"_padded;
  simdjson::ondemand::parser parser;
  simdjson::ondemand::document doc = parser.iterate(json);
  
  f(doc.at(0));
  return EXIT_SUCCESS;
}
```

The `simdjson.h` header does all of the work for you !!!

**Reply by longod** [reply]
Yes, normally it can be solved by including `simdjson.h` as you suggested.
What I wanted to ask more specifically is as follows:

my_api.h
```cpp
#pragma once

// need forward declaration or #include "simdjson.h"

void do_something(simdjson::ondemand::value& v);
```

my_api.cpp
```cpp
#include "my_api.h"
#include "simdjson.h"

void do_something(simdjson::ondemand::value& v) {
    // do something
}
```

main.cpp

```cpp
#include "my_api.h"
#include "simdjson.h"

int main() {
    simdjson::padded_string json = R"( [ 233897314173811950000 ])"_padded;
    simdjson::ondemand::parser parser;
    simdjson::ondemand::document doc = parser.iterate(json);
  
    f(doc.at(0));
    return 0;
}

```
In this situation, `my_api.h` requires either a **forward declaration** or the include of `simdjson.h`.

The easy solution is to include `simdjson.h` in `my_api.h`, but I would like to reduce the number of includes by forward declarations to prevent them from being included in too many files and increasing compile time. So I was trying forward declarations like https://github.com/simdjson/simdjson/discussions/2097#discussion-5968360, but it didn't work.

**Reply by lemire** [reply]
The performance of the On Demand approach depends critically on the ability of the compiler to inline function calls. In your scenario, how do you insure that your calls to `do_something` can be inlined (for performance)?

**Reply by longod** [reply]
I see. In other words, should I consider using it in a way that prioritizes runtime performance?

**Reply by lemire** [reply]
Small functions, functions that process one value, should typically be inlined. It can make a big difference.

You might be a Visual Studio user, people concerned with build times are often Visual Studio users. Visual Studio builds very slowly. But if you switch to ClangCL, you might see significant build time improvements. 

E.g., you might be able to double your build speed, and improve the runtime performance:
https://lemire.me/blog/2023/02/27/visual-studio-versus-clangcl/

**Reply by longod** [reply]
Thanks for the very interesting information.
Yes, I am a Visual Studio user. It seems to me that ClangCL is worth a try.

**Reply by lemire** [reply]
Chrome, Firefox and many triple-A games under Windows are built with ClangCL.
